agent/consul/cluster_endpoint.go

```diff
@@ -0,0 +1,298 @@
+		}
+
+		// generate the accessor id
+		accessor, err := lib.GenerateUUID(c.srv.checkTokenUUID)
+		if err != nil {
+			return err
+		}
+		// generate the secret id
+		secret, err := lib.GenerateUUID(c.srv.checkTokenUUID)
+		if err != nil {
+			return err
+		}
+
+		// set up the token
+		token := structs.ACLToken{
+			AccessorID:  accessor,
+			SecretID:    secret,
+			Description: fmt.Sprintf("Auto Config Token for Node %q", opts.NodeName),
+			CreateTime:  time.Now(),
+			Local:       true,
+			NodeIdentities: []*structs.ACLNodeIdentity{
+				{
+					NodeName:   opts.NodeName,
+					Datacenter: c.srv.config.Datacenter,
+				},
+			},
+			EnterpriseMeta: *structs.DefaultEnterpriseMeta(),
+		}
+
+		req := structs.ACLTokenBatchSetRequest{
+			Tokens: structs.ACLTokens{&token},
+			CAS:    false,
+		}
+
+		// perform the request to mint the new token
+		if _, err := c.srv.raftApplyMsgpack(structs.ACLTokenSetRequestType, &req); err != nil {
+			return err
+		}
+
+		acl.Tokens = &config.ACLTokens{Agent: secret}
+	}
+
+	conf.ACL = acl
+	return nil
+}
+
+// updateJoinAddressesInConfig determines the correct gossip endpoints that clients should
+// be connecting to for joining the cluster based on the segment given in the opts parameter.
+func (c *Cluster) updateJoinAddressesInConfig(opts AutoConfigOptions, conf *config.Config) error {
+	members, err := c.srv.LANSegmentMembers(opts.SegmentName)
+	if err != nil {
+		return err
+	}
+
+	var joinAddrs []string
+	for _, m := range members {
+		if ok, _ := metadata.IsConsulServer(m); ok {
+			serfAddr := net.TCPAddr{IP: m.Addr, Port: int(m.Port)}
+			joinAddrs = append(joinAddrs, serfAddr.String())
+		}
+	}
+
+	conf.RetryJoinLAN = joinAddrs
+	return nil
+}
+
+// updateGossipEncryptionInConfig will populate the gossip encryption configuration settings
+func (c *Cluster) updateGossipEncryptionInConfig(_ AutoConfigOptions, conf *config.Config) error {
+	// Add gossip encryption settings if there is any key loaded
+	memberlistConfig := c.srv.config.SerfLANConfig.MemberlistConfig
+	if lanKeyring := memberlistConfig.Keyring; lanKeyring != nil {
+		pk := lanKeyring.GetPrimaryKey()
+		if len(pk) > 0 {
+			conf.EncryptKey = base64.StdEncoding.EncodeToString(pk)
+		}
+
+		conf.EncryptVerifyIncoming = memberlistConfig.GossipVerifyIncoming
+		conf.EncryptVerifyOutgoing = memberlistConfig.GossipVerifyOutgoing
+	}
+
+	return nil
+}
+
+// updateTLSSettingsInConfig will populate the TLS configuration settings but will not
+// populate leaf or ca certficiates.
+func (c *Cluster) updateTLSSettingsInConfig(_ AutoConfigOptions, conf *config.Config) error {
+	// add in TLS configuration
+	conf.VerifyServerHostname = c.srv.tlsConfigurator.VerifyServerHostname()
+	base := c.srv.tlsConfigurator.Base()
+	conf.VerifyOutgoing = base.VerifyOutgoing
+	conf.TLSMinVersion = base.TLSMinVersion
+	conf.TLSPreferServerCipherSuites = base.PreferServerCipherSuites
```

Comment:
I think the CA should be transmitted as well. Either in here or in `updateTLSCertificatesInConfig`. I am not 💯 about all the different scenarios, but it would be nice to have on the client side.

Reply:
Right now thats happening via AutoEncrypt. A future update will take this from using auto encrypt to pushing down all the things.

Reply:
But what about the cases when AutoEncrypt is not enabled, because AutoConfig does not imply that, right?

Reply:
If AutoEncrypt is not enabled then with the current implementation the client will not get TLS certificates. Depending on the value of verify_incoming_rpc on the servers it may or may not work.